File: tnzhou/57.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        int current_left = -1;
        int current_right = -1;
        vector<Interval> result;
        bool inserted_new = false;
        if (intervals.size() == 0) {
            result.push_back(newInterval);
            return result;
        }
        int start_pos = 1;
        current_left = intervals[0].start;
        current_right = intervals[0].end;
        if (newInterval.start < current_left) {
            current_left = newInterval.start;
            current_right = newInterval.end;
            inserted_new = true;
            start_pos = 0;
        }
        for (int i = start_pos; i < intervals.size(); i ++) {
            Interval itv = intervals[i];
            if (!inserted_new && newInterval.start <= itv.start) {
                itv = newInterval;
                inserted_new = true;
                i--;
            }
            if (itv.start <= current_right && itv.end >= current_left) {
                current_left = min(itv.start, current_left);
                current_right = max(itv.end, current_right);
            } else {
                result.push_back(Interval(current_left, current_right));
                current_left = itv.start;
                current_right = itv.end;
            }
        }
        if (!inserted_new) {
            Interval itv = newInterval;
            if (itv.start <= current_right && itv.end >= current_left) {
                current_left = min(itv.start, current_left);
                current_right = max(itv.end, current_right);
            } else {
                result.push_back(Interval(current_left, current_right));
                current_left = itv.start;
                current_right = itv.end;
            }
        }
        result.push_back(Interval(current_left, current_right));
        return result;
    }
};

File: zemiao/57.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        res = []
        newV = newInterval
        inserted = False
        for v in intervals:
          if inserted or v.end < newV.start:
            res.append(v)
          elif v.start > newV.end:
            res.append(newV)
            res.append(v)
            inserted = True
          else:
            newV = Interval(min(v.start,newV.start),max(v.end,newV.end))
        if not inserted:
          res.append(newV)
        return res